src/main/java/org/jsoup/helper/CookieUtil.java

```
@@ -0,0 +1,88 @@
+        // Request key/val cookies. LinkedHashSet used to preserve order, as cookie store will return most specific path first
+        Set<String> cookieSet = requestCookieSet(req);
+        Set<String> cookies2 = null;
+
+        // stored:
+        Map<String, List<String>> storedCookies = req.cookieManager().get(asUri(req.url), EmptyRequestHeaders);
+        for (Map.Entry<String, List<String>> entry : storedCookies.entrySet()) {
+            // might be Cookie: name=value; name=value\nCookie2: name=value; name=value
+            List<String> cookies = entry.getValue(); // these will be name=val
+            if (cookies == null || cookies.size() == 0) // the cookie store often returns just an empty "Cookie" key, no val
+                continue;
+
+            String key = entry.getKey(); // Cookie or Cookie2
+            Set<String> set;
+            if (CookieName.equals(key))
+                set = cookieSet;
+            else if (Cookie2Name.equals(key)) {
+                set = new HashSet<>();
+                cookies2 = set;
+            } else {
+                continue; // unexpected header key
+            }
+            set.addAll(cookies);
+        }
+
+        if (cookieSet.size() > 0)
+            con.addRequestProperty(CookieName, StringUtil.join(cookieSet, Sep));
+        if (cookies2 != null && cookies2.size() > 0)
+            con.addRequestProperty(Cookie2Name, StringUtil.join(cookies2, Sep));
+    }
+
+    private static LinkedHashSet<String> requestCookieSet(Connection.Request req) {
+        LinkedHashSet<String> set = new LinkedHashSet<>();
+        // req cookies are the wildcard key/val cookies (no domain, path, etc)
+        for (Map.Entry<String, String> cookie : req.cookies().entrySet()) {
+            set.add(cookie.getKey() + "=" + cookie.getValue());
+        }
+        return set;
+    }
+
+    static URI asUri(URL url) throws IOException {
+        try {
+            return url.toURI();
+        } catch (URISyntaxException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

---

Updated, thanks